plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,211 @@
+ * limitations under the License.
+ */
+package io.trino.plugin.password.ldap;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.CharMatcher;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+import io.airlift.log.Logger;
+import io.trino.collect.cache.EvictableCacheBuilder;
+import io.trino.plugin.base.ldap.LdapClient;
+import io.trino.plugin.base.ldap.LdapQuery;
+import io.trino.spi.security.AccessDeniedException;
+import io.trino.spi.security.GroupProvider;
+
+import javax.inject.Inject;
+import javax.naming.NamingException;
+
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class LdapGroupProvider
+        implements GroupProvider
+{
+    private static final Logger log = Logger.get(LdapGroupProvider.class);
+    private static final Pattern EXTRACT_GROUP_PATTERN = Pattern.compile("(?i)cn=([^,]*),(.*)$");
+    static final String GROUP_SEARCH_FILTER_PREFIX = "member=";
+    private static final CharMatcher SPECIAL_CHARACTERS = CharMatcher.anyOf(",=+<>#;*()\"\\\u0000");
+    private static final CharMatcher WHITESPACE = CharMatcher.anyOf(" \r");
+
+    private final LdapClient client;
+
+    private final Optional<String> userBaseDistinguishedName;
+    private final Optional<String> groupBaseDistinguishedName;
+    private final Optional<String> bindDistinguishedName;
+    private final Optional<String> bindPassword;
+    private final boolean allowUserNotExist;
+
+    private final LoadingCache<String, Set<String>> groupCache;
+
+    @Inject
+    public LdapGroupProvider(LdapClient client, LdapGroupProviderConfig ldapConfig)
+    {
+        this.client = requireNonNull(client, "client is null");
+        requireNonNull(ldapConfig, "ldapConfig is null");
+
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.groupBaseDistinguishedName = Optional.ofNullable(ldapConfig.getGroupBaseDistinguishedName());
+        this.bindDistinguishedName = Optional.ofNullable(ldapConfig.getBindDistinguishedName());
+        this.bindPassword = Optional.ofNullable(ldapConfig.getBindPassword());
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+
+        checkArgument(bindDistinguishedName.isPresent() == bindPassword.isPresent(), "Both bind distinguished name and bind password must be provided");
```

Comment:
Can we make them non-Optional and maybe annotate with `@NotNull` for the getter method ?